components/compositing/compositor.rs

```diff
@@ -952,6 +955,62 @@ impl IOCompositor {
         }
     }
 
+    fn maybe_send_awaiting_display_list(&mut self) {
+        let (built_display_list, display_list_info, webview_id, done) = {
+            let Some((_, display_list_info, _, _)) = &mut self.awaiting_display_list else {
+                return;
+            };
+            for (k, v) in &self.image_epochs {
```

Comment:
I don't think this will work as expected, because you are sending the epoch for an image from script to the painter and then to the compositor, but concurrently you are sending the display list from script: so when the display list is received, the epoch could be none. 
You could try to fix it by sending the epoch from script directly to the compositor, before sending the canvas message.

Reply:
We store received epoch for each key on add/update image commands. When we receive DL we wait until all epochs are new enough (we do rechecks after each add/update image commands), so there should be no problem with epoch could be none, as that would just mean DL will be applied later on some add/update image command.

The only problem that currently webgpu has is that it generates new display lists with new images faster then resolving them, so we never apply any display list, but that should be fixed by compositor driven update the rendering.

Reply:
No I disagree, because despite missing the initial image update(when it would still be none when the display list is received), you could also get into a situation where you always keep missing the epoch.

Basically, each time script increments the epoch and sends it to canvas, by the time the compositor receives the update image message, it could be already out of date. 

I now see that the expected epoch is in the display list info, so I think you should restructure the loop here, so that it is driven by the stuff that is in the display list, not the stuff stored in the compositor.